torch/csrc/jit/python/init.cpp

```diff
@@ -897,13 +898,65 @@ void initJITBindings(PyObject* module) {
             docstring << "  " << op->schema() << "\n";
           }
 
-          return py::cpp_function(
-              [operations](py::args args, py::kwargs kwargs) {
+          // To support __torch_function__ override mechanism, `func` needs to
+          // provide a pointer, which points to the original method, i.e.,
+          // itself. py::cpp_function doesn't support specifying the function
+          // signature without real implementation, so we we use a trick here.
+          // First, we create a function holder pointing to an empty
+          // cpp_function. Then we use a reference as the alias of the empty
+          // function. Then we define the real returned function, and update
+          // the value of the function pointed by the holder using the alias.
+          // With this trick, we are able to reference self inside the returned
+          // function.
+          auto func_holder = std::make_unique<py::cpp_function>();
+          py::cpp_function& func_ref = *func_holder;
+          auto func = py::cpp_function(
+              [operations, symbol, self_ptr = std::move(func_holder)](
+                  py::args args, py::kwargs kwargs) {
+                std::vector<py::handle> overloaded_args;
+                for (size_t i = 0; i < args.size(); ++i) {
+                  extract_torch_function_overloaded_arg(
+                      args[i].ptr(), overloaded_args);
+                }
+                // NB: for kwargs, we cannot guarantee the order of appending
+                // is the same as the argument order in operator's schema.
+                // This is suboptimal, but should be fine. Later when we have
+                // better schema matching and argument parsering, we could
+                // match the operator in `operations` first, then the order will
+                // be guaranteed.
+                for (auto item : kwargs) {
+                  extract_torch_function_overloaded_arg(
+                      item.second.ptr(), overloaded_args);
+                }
+                if (overloaded_args.size() > 0) {
+                  std::vector<py::object> overloaded_types;
+                  overloaded_types.reserve(overloaded_args.size());
+                  for (auto& oarg : overloaded_args) {
+                    overloaded_types.push_back(
+                        py::reinterpret_borrow<py::object>(
+                            (PyObject*)Py_TYPE(oarg.ptr())));
+                  }
+                  py::tuple py_types = py::cast(overloaded_types);
```

Comment:
Efficiency nit: it would be more efficient to directly put the overloaded types into the python tuple directly, rather than going through a vector

Reply:
Do we have a way to append to tuple?

Reply:
You cannot append to a tuple but in the raw Python API you preallocate a tuple of the desired size and then directly set elements at each index. I don't know if pybind11 gives a nice way to do this, you'd have to check their docs.

Reply:
just realized python tuple is immutable.